### PR TITLE
Add tests for field casing, duplicate field ordering, and unknown field errors when performing JSON deserialization

### PR DESCRIPTION
**Pull request by hudlow**

### CODE REVIEW

File: packages/protobuf-test/src/json.test.ts [hunk 1]
```diff
@@ -833,6 +857,56 @@ describe("JsonWriteOptions", () => {
...
+        '{ "singularStringField": "b" }',
+      );
+
+      expect(a.singularStringField).toBe("a");
+      expect(b.singularStringField).toBe("b");
+    });
+  });
+  describe("duplicate fields", () => {
+    // This depends on the ECMA-262-defined JSON.parse() behavior, which is itself
+    // specified to choose the last field value for a duplicate field.
```
Comment (timostamm) [inline packages/protobuf-test/src/json.test.ts:878]: This behavior has been discussed in https://github.com/bufbuild/protobuf-es/issues/1038. I'm not sure these tests are particularly useful, because they basically test JSON.parse 🤔 

I just noticed that the [ProtoJSON documentation](https://protobuf.dev/programming-guides/json/) was recently [updated](https://github.com/protocolbuffers/protocolbuffers.github.io/commit/c18c8a72de22b99eaf536eb1f961371510159714#diff-98c94e5ab37db71fd002eff0d9cb2cfae9b0058ea00b412b466312ad819029d7) with the following:

> Multiple values for
singular fields (using duplicate or equivalent JSON keys) are accepted and the
last value is retained, as with binary format parsing. Note that not all
protobuf JSON parser implementations are conformant, and some nonconformant
implementations may reject duplicate keys instead.

With JSON.parse, the last property also wins for non-singular fields:

```ts
    const a = fromJsonString(
      proto3_ts.Proto3MessageSchema,
      '{ "repeatedStringField": ["b"], "repeatedStringField": ["a"] }',
    );
    const b = fromJsonString(
      proto3_ts.Proto3MessageSchema,
      '{ "repeatedStringField": ["a"], "repeatedStringField": ["b"] }',
    );

    expect(a.repeatedStringField).toStrictEqual(["a"]);
    expect(b.repeatedStringField).toStrictEqual(["b"]);
```

Quite different to the binary format, where repeated values accumulate...

Comment (hudlow) [reply packages/protobuf-test/src/json.test.ts:878]: > This behavior has been discussed in https://github.com/bufbuild/protobuf-es/issues/1038.

I commented there.

> I'm not sure these tests are particularly useful, because they basically test JSON.parse 🤔

That's right, and in that sense they are certainly not _unit_ tests. I think they are worthwhile for completeness, but I'm certainly open to being convinced they're not necessary in this testing context.

> I just noticed that the [ProtoJSON documentation](https://protobuf.dev/programming-guides/json/) was recently [updated](https://github.com/protocolbuffers/protocolbuffers.github.io/commit/c18c8a72de22b99eaf536eb1f961371510159714#diff-98c94e5ab37db71fd002eff0d9cb2cfae9b0058ea00b412b466312ad819029d7)

Yeah, I read that and it's part of what prompted me to define the as-is behavior as the "correct" baseline. I was actually doing that _in preparation_ for filing what I now know would have been a dupe of #1038. 🤪

> Quite different to the binary format, where repeated values accumulate...

I think you mean to say that ProtoJSON arrays are different from the binary format where array elements can be serialized non-contiguously. 😉

File: packages/protobuf-test/src/json.test.ts [hunk 2]
```diff
@@ -833,6 +857,56 @@ describe("JsonWriteOptions", () => {
   });
 });
 
+describe("parsing funny JSON", () => {
+  describe("field casing", () => {
+    test("is case agnostic", () => {
```
[the remark this answers is not in the document]
Comment (hudlow) [reply packages/protobuf-test/src/json.test.ts]: Removed.

File: packages/protobuf-test/src/json.test.ts (hunk 1 above)
Comment (hudlow) [reply packages/protobuf-test/src/json.test.ts:863]: > Arrays are non-contiguous in the binary format - well said. It just occurred to me that behavior is also different with singular message fields. They can occur multiple times in the binary format, which will [merge](https://protobuf.com/docs/language-spec#merging-protobuf-messages) them, unlike ProtoJSON.

Interesting; I don't think I comprehend the benefit of this for a wire format. (And, as with ignoring duplicated JSON fields, I _do_ see the drawbacks.)

There are also some interesting things around the edges of the merge algorithm, if I'm reading it right:

- Map values for the same key don't get recursively merged even when they are a mergeable type (a repeated field, a message, or another map). This means (I think) that you couldn't replicate the merge logic in two ProtoJSON representations without considering the underlying types. (Probably not a huge deal, but a small blow to the self-describing benefits of JSON serialization.)
- Like JSON, Protobuf is missing a set type, so there's no way to signal that a merged repeated field should be deduplicated.

Comment (hudlow) [reply packages/protobuf-test/src/json.test.ts:863]: > I don't think I comprehend the benefit of this for a wire format.

Is the benefit that you can stream a Protobuf message and update a field along the way?

**Comment by hudlow** [comment]
Currently blocked by https://github.com/protocolbuffers/protobuf/pull/21873 which is ongoing.

**Comment by hudlow** [comment]
Should be up-to-date and ready for another look.
